CONTRIBUTING.txt

```diff
@@ -0,0 +1,99 @@
+
+
+0. Status
+
+FreeCAD is in a transition period. The following are to be regarded as GUIDELINES for contribution submission and acceptance. For historical reasons, the actual process MAY diverge from this process during the transition.  Such deviations SHOULD be noted and discussed whenever possible.
+
+1. Goals
+
+    The FreeCAD Contribution Process is expressed here with the following specific goals in mind:
+
+    1. To provide transparency and fairness in the contribution process.
+    2. To allow useful contributions to be included as quickly as possible.
+    3. To preserve and improve the code quality while encouraging appropriate experimentation and risk-taking.
+    4. To minimize dependence on individual Contributors by encouraging a large pool of active Contributors.
+    5. To be inclusive of many viewpoints and to harness a diverse set of skills.
+    6. To provide an encouraging environment where Contributors learn and improve their skills.
+    7. To protect the free and open nature of the FreeCAD project.
+
+2. Fundamentals
+
+    1. FreeCAD uses the git distributed revision control system.
+    2. Source code for the main application and related subprojects is hosted on github.com in the FreeCAD organization, hereafter referred to as 'the Platform'.
+    3. Problems are discrete, well-defined limitations or bugs.
+    4. FreeCAD uses the Platform's issue-tracking system to track problems.
+    5. Patches are sets of code changes that resolve a single problem.
+    6. FreeCAD uses the Pull Request workflow for evaluating and accepting patches.
+
+3. Roles
+    1. "User": An member of the wider FreeCAD community who uses the software.
+    2. "Contributor":  A person who submits a patch that resolves a previously identified problem. Contributors do not have commit access to the repository unless they are also Maintainers. Everyone, without distinction or discrimination, SHALL have an equal right to become a Contributor.
+    3. "Maintainer": A person who merges patches. Maintainers may or may not be Contributors. Their role is to enforce the process. Maintainers have commit access to the repository.
+    4. "Administrator": Administrators have additional authority to maintain the list of designated Maintainers.
+
+4. Licensing, Ownership, and Credit
+    1. FreeCAD is distributed under the Lesser General Public License, version 2, or superior (LGPL2+).  Additional details can be found in the LICENSE file.
+    2. All contributions to FreeCAD MUST use a compatible license.
+    3. All contributions are owned by their authors unless assigned to another.
+    4. FreeCAD does not have a mandatory copyright assignment policy.
+    5. A Contributor who wishes to be identified in the Credits section of the application "About" dialog is responsible for identifying themselves appropriately by notifying an Administrator or Maintainer.
+
+5. Contribution Requirements
+
+    1. Patches are submitted in the form of Pull Requests (PR).
```

Comment:
Yes. And even goo further "A patch shall be submitted in the form of a Pull Request. Although I don't know why persist "patch". I suspect it is in the C4 document, but that may have been originally written before git. I believe "patch" means something different nowadays.

Reply:
> Although I don't know why persist "patch". I suspect it is in the C4 document, but that may have been originally written before git.

"Patch" is defined in the document. It stands for the changeset proposed to the project code (i.e. the "content").
It's IMO correct, but what would you propose as an alternative ?

Reply:
"patch" is used in the document. In the days before git "patch' and "diff" were the tools of trade. Since git by far the most common unit of posting is a PR. I suspect the C4 document from which this text has been lifted (oops, did I say that?) may have been started/written years ago, possible pre git, and reflects old ways, with a fix to bring it up to date. Github has only risen to supremacy in the last several years. We have the luxury of a fresh start, and IMO do not need to start with obsolete terms. Nowadays the unit of change is clearly the PR.

Reply:
You're mixing up content and tool.
'Patch' is the content. 'PR' is the tool (mechanism) that allows to integrate patches into the codebase. 'Using GH PR' is the process chosen by the project.

A patch is a patch. Not because a tool that was used in older processes has been named like this we shall stop using this word.

Reply:
Yeah I get that Pull Request and Patch are somewhat synonomous, but I do not see "patch" come up very often when working with git/GitHub.

Reply:
I still see it used regularly on either our bugzilla or on GitHub when talking about fixes, package updates, etc.

Reply:
OK, I take your word for that, but is that because it is the terminology used interacting with git, or just old habits. FC still has a Bugzilla?

Reply:
Not FC, but Gentoo still has bugzilla.
AFAICS the term patch was first, from patching, i.e. repairing something / code. From these the name of the tool patch arose. The tool is still used as of today. On Gentoo we use it in the package build process to apply patches, which often come from a PR on GitHub through the means of the `git format-patch` command.
Only because the tool is ancient, doesn't mean the original meaning of patching something doesn't apply anymore. I see no sense not to use this wording anymore. This has nothing to do with old habits, but because it's a precise description of what is going on.

Reply:
I really can't get it. 'Patch' is just the correct word (IMO) to define what is expected to be the content of a PR : https://en.m.wikipedia.org/wiki/Patch_(computing)
Notice 'PR' isn't git terminology but GH one.

Reply:
I struggled with this.  OpenBrain's comment about content vs tool helped.
The C4 document was written to be platform agnostic.  When I started this rewrite, I chose to just say 'Github' since that's what we use.  People wanted to keep it more agnostic so I reverted.
The problem is that the process wants a single set of changes (ie "Patch") to correspond to a single issue (ie "Problem") but the PR may contain multiple commits.  So which one is a patch, the commit or the PR?  
IMHO, it's the PR but we allow multiple commits within the PR to give the contributor flexibility and to keep the change process cleaner.  

If someone is proposing a text change, please do so.  Otherwise, I'll resolve this in the next day or so.

Reply:
> If patch is contentious, "contribution" is used elsewhere in the document.
> 
>     1. Contributions shall be submitted in the form of Pull Requests (PR).

I like this formulation. It abstracts from the terms patch and PR yet is precide enough and hopefully well understood by all parties.

Reply:
Patch is a generic term, both are patches

Reply:
> I'd say the commit is the patch.

Actually not really. 'Patch' is neither the commit(s), nor the PR, nor anything else. 'Patch' is the changeset to be applied to the code. It can be provided as commit(s), as diff file, or any other mean. ;)



> If patch is contentious, "contribution" is used elsewhere in the document.
> 
> 1. Contributions shall be submitted in the form of Pull Requests (PR).

Looks wise. Especially it includes contributions that hardly classified as 'patch', such as providing a new icon for example.

Reply:
Done.  "patch' is gone in favor of "contribution"